### PR TITLE
Authentication

### LINKED ISSUE #1: Implement authentication in backend
- [x] Create Mongoose user schema
- [x] Create GraphQL users schema
- [x] Create register mutation
- [x] Hash passwords
- [x] Create login query
- [x] Create JWT
- [x] Enable cookies to retrieve client jwt
- [x] Create auth middleware
- [x] Create logout query

### PR DESCRIPTION
**Pull request by auloin**
Quite a lot already.
Need a critical look at the folder structures.
Where should I put Business logic (like email validity check)? In the model? A custom middleware in between graphql & mongoose?

Closes #2 
Closes #1

### CODE REVIEW

File: backend/src/graphql/resolvers/queries.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
+const bcrypt = require('bcrypt');
+const jwt = require('jsonwebtoken');
+const User = require('../../models/user');
+
+module.exports = {
+  users: async () => (users = await User.find()),
```
Comment (MichielLeyman) [inline backend/src/graphql/resolvers/queries.js]: `users` is undefined. Did you want to return `User.find()`?

File: backend/src/graphql/resolvers/mutations.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
+const bcrypt = require('bcrypt');
+const User = require('../../models/user');
+
+module.exports = {
+  register: async (_, { user }) => {
+    // TODO Valid email format
```
Comment (MichielLeyman) [inline backend/src/graphql/resolvers/mutations.js]: Are you doing validation all at once in a separate PR?
Might be worth doing straight away

File: backend/src/index.js [hunk 3]
```diff
@@ -1,30 +1,53 @@
-const { ApolloServer, gql } = require('apollo-server');
+require('dotenv').config();
+const mongoose = require('mongoose');
+const express = require('express');
```
Comment (MichielLeyman) [inline backend/src/index.js]: Not listed in dependencies in `package.json`

File: backend/src/graphql/resolvers/mutations.js (hunk 2 above)
Comment (archived-m) [inline backend/src/graphql/resolvers/mutations.js]: Are you doing validation all at once in a separate PR?
Might be worth doing straight away

File: backend/src/graphql/resolvers/queries.js (hunk 1 above)
Comment (archived-m) [inline backend/src/graphql/resolvers/queries.js]: `users` is undefined. Did you want to return `User.find()`?

File: backend/src/index.js (hunk 3 above)
Comment (archived-m) [inline backend/src/index.js:4]: Not listed in dependencies in `package.json`

File: backend/src/models/bankDetails.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
+const mongoose = require('mongoose');
+
+const { Schema } = mongoose;
+
+const bankDetailsSchema = new Schema(
```
Comment (archived-m) [inline backend/src/models/bankDetails.js:5]: Is a constructor, bot the variable and filename should be capitalized

File: backend/src/models/user.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
+const mongoose = require('mongoose');
+const Address = require('./address');
+const BankDetails = require('./bankDetails');
+
+const { Schema } = mongoose;
+
+const userSchema = new Schema({
```
Comment (archived-m) [inline backend/src/models/user.js:9]: Is a constructor, both the variable and filename should be capitalized

File: backend/src/models/address.js [hunk 6]
```diff
@@ -0,0 +1,15 @@
+const mongoose = require('mongoose');
+
+const { Schema } = mongoose;
+
+const addressSchema = new Schema(
```
Comment (archived-m) [inline backend/src/models/address.js:5]: Is a constructor, bot the variable and filename should be capitalized

File: backend/src/models/bankDetails.js (hunk 4 above)
Comment (archived-m) [inline backend/src/models/bankDetails.js:5]: So right now there's one payment method, being a bank account?

File: backend/src/graphql/resolvers/queries.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+  login: async (_, { email, password }, { res }) => {
+    // lowerCase the email && check if the user exists
+    const user = await User.findOne({ email: email.toLowerCase() });
+    // compare provided password against stored password
+    let rightPassword = false;
+    if (user) {
+      rightPassword = await bcrypt.compare(password, user.password);
+    }
+    if (!user || !rightPassword) {
```
Comment (archived-m) [inline backend/src/graphql/resolvers/queries.js]: Maybe `try` `catch` here, if it errors at any point, the email or password is incorrect.

File: backend/src/index.js [hunk 8]
```diff
@@ -1,30 +1,53 @@
...
-  onHealthCheck: () => fetch('https://fourtonfish.com/hellosalut/?mode=auto'),
+  resolvers: {
+    Query,
+    Mutation
+  },
+  context: async ({ req, res }) => {
+    // Get the token
+    const { token } = req.cookies;
+    let user = null;
+    if (token) {
```
Comment (archived-m) [inline backend/src/index.js]: Can nest fewer ifs if we check for `!token` first and return a 401

File: backend/src/graphql/resolvers/queries.js (hunk 1 above)
Comment (archived-m) [inline backend/src/graphql/resolvers/queries.js]: Is this query for admins to get a listing of all users?

File: backend/src/graphql/resolvers/mutations.js [hunk 9]
```diff
@@ -0,0 +1,24 @@
+const bcrypt = require('bcrypt');
+const User = require('../../models/user');
+
+module.exports = {
+  register: async (_, { user }) => {
```
Comment (archived-m) [inline backend/src/graphql/resolvers/mutations.js]: Someone who freshly registers, can only create expenses right? If not, there needs to be an approve/reject mechanism

File: backend/src/graphql/resolvers/mutations.js [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+    // lowercase the email
+    const email = user.email.toLowerCase();
+    // Make sure any user has the email
+    const dupUser = await User.findOne({ email });
+    if (dupUser) {
+      throw new Error('An account with that email address already exists.');
+    }
+    // Hash Password
+    const password = await bcrypt.hash(user.password, 10);
+    // Register && Return the user
```
Comment (archived-m) [inline backend/src/graphql/resolvers/mutations.js]: TODO: send account confirmation email

**Review by archived-m — COMMENTED** [review]
Great work.
Maybe it would be good to think about mailing in here already - and set up a password reset flow while we're on the topic of authentication.

If admins already have a way to get all users, why is there no way to update a user (set their role or `active` state) or delete them altogether?

**Review by archived-m — CHANGES_REQUESTED** [review]
Also, please add a basic README on which server environment variables are required for the app to start.

File: backend/src/graphql/resolvers/queries.js (hunk 1 above)
Comment (auloin) [reply backend/src/graphql/resolvers/queries.js]: I'll add authorization later as well as roles

File: backend/src/graphql/resolvers/queries.js (hunk 7 above)
Comment (auloin) [reply backend/src/graphql/resolvers/queries.js]: There is no error handling yet, I'll add that later.

File: backend/src/index.js (hunk 3 above)
Comment (auloin) [reply backend/src/index.js:4]: Yes there were a need to switch to apollo-server-express

File: backend/src/index.js (hunk 8 above)
Comment (auloin) [reply backend/src/index.js]: This mean anybody can register or login. I'm thinking of extract this logic in a separate function.

File: backend/package.json [hunk 11]
```diff
@@ -11,11 +11,20 @@
...
+    "dotenv": "^7.0.0",
     "graphql": "^14.1.1",
+    "jsonwebtoken": "^8.5.1",
     "mongoose": "^5.4.19"
   },
   "devDependencies": {
+    "eslint": "^5.3.0",
+    "eslint-config-airbnb": "^17.1.0",
+    "eslint-plugin-import": "^2.16.0",
+    "eslint-plugin-jsx-a11y": "^6.2.1",
```
Comment (auloin) [inline backend/package.json:29]: Are react related plugins useful in the backend?

**Review by auloin — COMMENTED** [review]
> Also, please add a basic README on which server environment variables are required for the app to start.

I didn't see this one. I will make sure to add a .env.example file and explain all variables in the readme.

**Comment by auloin** [comment]
What do you think of using directives for validation & authorization? A working example of directive here 
9d0dcaa.

File: backend/package.json (hunk 11 above)
Comment (archived-m) [reply backend/package.json:29]: I think they're required for the airbnb style guide

**Review by archived-m — APPROVED** [review]

**Pull request merged**